### ISSUE TITLE
delegate/function overload doesn't work anymore

### ISSUE THREAD

**Issue by disarray2077** [beefytech/Beef#1592]
Looks like the following code broke at some point:
```cs
public static void TestOverload<TProc>(TProc proc)
	where TProc : class, delegate void()
{
	Console.WriteLine("1");
}

public static void TestOverload<TProc>(TProc proc)
	where TProc : struct, function void()
{
	Console.WriteLine("2");
}

public static void Main(String[] args)
{
	TestOverload/*<function void()>*/(() => {}); // ERROR: The type 'method reference test6.Program.Main.anon()' must be a reference type in order to use it as parameter 'TProc' for 'test6.Program.TestOverload<method reference test6.Program.Main.anon()>(method reference test6.Program.Main.anon() proc)'

}
```
Tested with https://github.com/beefytech/Beef/commit/5268e103e9bac56211586febcffd1fecdca25d61

**Comment by bfiete** [comment]
What happened before was that #1 was incorrectly not failing on the `class` constraint (method references are not classes) so it was binding to #1.

#2 probably wasn't even getting called at all, even for functions (I think?). Either way, you really only need:

```
public static void TestOverload<TProc>(TProc proc)
	where TProc : delegate void()
{
	Console.WriteLine("1");
}

public static void TestOverload(function void() proc)
{
	Console.WriteLine("2");
}
```

You only need one version for method references to bind to, and they will only bind to delegate constraints anyway. 
If you need another version that _disallows_ method references then you can add a variant of #1 except with the `class` constraint added.

Or is there something else I'm missing?

**Comment by disarray2077** [comment]
>#2 probably wasn't even getting called at all, even for functions (I think?)

Oh... You're right, that's weird. Now I'm also not sure if it was working before lol. The only calls in my code were like this, which used to work with the function constraint:
```cs
function void() f = () => {};
TestOverload(f);
```

Anyway, I found one issue when the `struct` and `delegate` constraints are together:
In the code below looks like the compiler picks the `TestOverload` method but the compilation fails because the constraint doesn't have parameters but it expects the `this` parameter:
```cs
public static void TestOverload<TProc>(TProc proc)
	where TProc : struct, delegate void()
{
	proc();
	Console.WriteLine("2");
}

public void VoidClassMethod()
{
}

public static void Main(String[] args)
{
	function void(Program this) f = => VoidClassMethod;
	TestOverload(f);
}
```

**Comment by bfiete** [comment]
Fixed at f8e57ee5ffca4dbc4a722351785f1d7f0abf3d3f
